src/main/java/de/tum/in/www1/artemis/service/StudentExamService.java

```diff
@@ -67,6 +88,65 @@ public StudentExam findOneWithExercisesByUserIdAndExamId(Long userId, Long examI
         return studentExamRepository.findWithExercisesByUserIdAndExamId(userId, examId);
     }
 
+    /**
+     * submit StudentExam and uses submissions as final submissions if studentExam is not yet submitted
+     * and if it was submitted after exam startDate and before individual endDate + gracePeriod
+     *
+     * @param studentExam latest studentExam object which will be submitted (final submission)
+     */
+    public void submitStudentExam(StudentExam studentExam) {
+        log.debug("Submit student exam with id {}", studentExam.getId());
+        // checks if student exam is already marked as submitted
+        if (studentExam.isSubmitted()) {
+            throw new IllegalStateException("StudentExam is already marked as submitted");
+        }
+
+        // gets individual exam end or exam.endDate if individual cannot be calculated
+        ZonedDateTime examEndDate = studentExam.getExam().getStartDate() != null && studentExam.getWorkingTime() != null
+                ? studentExam.getExam().getStartDate().plusSeconds(studentExam.getWorkingTime())
+                : studentExam.getExam().getEndDate();
+
+        // checks if student exam is live (after start date, before end date + grace period)
+        if ((studentExam.getExam().getStartDate() != null && !ZonedDateTime.now().isAfter(studentExam.getExam().getStartDate()))
+                || (examEndDate != null && !(ZonedDateTime.now().isBefore(examEndDate.plusSeconds(studentExam.getExam().getGracePeriod()))))) {
+            throw new IllegalStateException("StudentExam cannot be marked as submitted, because it is not invoked between start and end of exam");
+        }
+
+        studentExam.getExercises().forEach(exercise -> {
+            // if exercise is either QuizExercise, TextExercise or ModelingExercise and exactly one participation exists
+            if (exercise.getStudentParticipations() != null && exercise.getStudentParticipations().size() == 1) {
+                exercise.getStudentParticipations().forEach(studentParticipation -> {
+                    // if exactly one submission exists we save the submission
+                    if (studentParticipation.getSubmissions() != null && studentParticipation.getSubmissions().size() == 1) {
+                        studentParticipation.setExercise(exercise);
+                        studentParticipation.getSubmissions().forEach(submission -> {
+                            submission.setParticipation(studentParticipation);
+                            submission.submissionDate(ZonedDateTime.now());
```

Comment:
Security: we should compare the submission with the existing one, so that no wrong data can be injected here (e.g. a submission with a different ID that overrides another submission of another user)

Reply:
See https://github.com/ls1intum/Artemis/pull/1867/commits/f73351dba1cdc0c02a9f7fe25a6bcebae281fff5